File: apps/desktop/packages/renderer/src/lib/queries/agent-runtime.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import type {
  AgentRuntimeSnapshot,
  AuthProfileLoginResult,
  AuthProfileLogoutResult,
  DaemonAgentRuntimeAuthLoginParams,
  DaemonAgentRuntimeAuthLogoutParams,
  DaemonAgentRuntimePatchProfileParams,
  DaemonAgentRuntimeSelectProfileParams,
  DaemonAgentRuntimeSetExtensionEnabledParams,
  DaemonAgentRuntimeTestLocalEndpointParams,
  LocalModelEndpointTestResult,
} from "@taugentic/desktop-shared";

import {
  getAgentRuntime,
  loginAgentRuntimeAuthProfile,
  logoutAgentRuntimeAuthProfile,
  patchAgentRuntimeProfile,
  selectAgentRuntimeProfile,
  setAgentRuntimeExtensionEnabled,
  testLocalModelEndpoint,
} from "@/lib/ipc/api";

import { agentRuntimeRootKey, queryKeys } from "./keys";
import { SESSION_QUERY_POLL_INTERVAL_MS, type SessionQueryView } from "./session-queries";

export function useAgentRuntimeQuery(): SessionQueryView<AgentRuntimeSnapshot> {
  const query = useQuery({
    queryKey: queryKeys.agentRuntime.snapshot,
    queryFn: () => getAgentRuntime(),
    refetchInterval: SESSION_QUERY_POLL_INTERVAL_MS,
  });
  return {
    data: query.data,
    error: query.error,
    isLoading: query.isLoading,
    isFetching: query.isFetching,
    refetch: query.refetch,
  };
}

export function useSelectAgentRuntimeProfileMutation() {
  const qc = useQueryClient();
  return useMutation<AgentRuntimeSnapshot, Error, DaemonAgentRuntimeSelectProfileParams>({
    mutationFn: (params) => selectAgentRuntimeProfile(params),
    onSuccess: (snapshot) => {
      qc.setQueryData(queryKeys.agentRuntime.snapshot, snapshot);
      void qc.invalidateQueries({ queryKey: agentRuntimeRootKey });
    },
  });
}

export function usePatchAgentRuntimeProfileMutation() {
  const qc = useQueryClient();
  return useMutation<AgentRuntimeSnapshot, Error, DaemonAgentRuntimePatchProfileParams>({
    mutationFn: (params) => patchAgentRuntimeProfile(params),
    onSuccess: (snapshot) => {
      qc.setQueryData(queryKeys.agentRuntime.snapshot, snapshot);
      void qc.invalidateQueries({ queryKey: agentRuntimeRootKey });
    },
  });
}

export function useLoginAgentRuntimeAuthProfileMutation() {
  const qc = useQueryClient();
  return useMutation<AuthProfileLoginResult, Error, DaemonAgentRuntimeAuthLoginParams>({
    mutationFn: (params) => loginAgentRuntimeAuthProfile(params),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: agentRuntimeRootKey });
    },
  });
}

export function useLogoutAgentRuntimeAuthProfileMutation() {
  const qc = useQueryClient();
  return useMutation<AuthProfileLogoutResult, Error, DaemonAgentRuntimeAuthLogoutParams>({
    mutationFn: (params) => logoutAgentRuntimeAuthProfile(params),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: agentRuntimeRootKey });
    },
  });
}

export function useSetAgentRuntimeExtensionEnabledMutation() {
  const qc = useQueryClient();
  return useMutation<AgentRuntimeSnapshot, Error, DaemonAgentRuntimeSetExtensionEnabledParams>({
    mutationFn: (params) => setAgentRuntimeExtensionEnabled(params),
    onSuccess: (snapshot) => {
      qc.setQueryData(queryKeys.agentRuntime.snapshot, snapshot);
      void qc.invalidateQueries({ queryKey: agentRuntimeRootKey });
    },
  });
}

export function useTestLocalModelEndpointMutation() {
  return useMutation<
    LocalModelEndpointTestResult,
    Error,
    DaemonAgentRuntimeTestLocalEndpointParams
  >({
    mutationFn: (params) => testLocalModelEndpoint(params),
  });
}

File: crates/ta-orchestrator/src/orchestration/app/agent_runtime.rs
use ta_store::PersistenceStore;

use crate::{
    AuthProfileLoginResult, AuthProfileLogoutResult, DaemonAgentRuntimeAuthLoginCompleteParams,
    DaemonAgentRuntimeAuthLoginParams, DaemonAgentRuntimeAuthLogoutParams,
    DaemonAgentRuntimeAuthProfilePreferencesSetParams, DaemonAgentRuntimePatchProfileParams,
    DaemonAgentRuntimeSetExtensionEnabledParams, GetAgentRuntimeQuery,
};

use super::{AppService, AppServiceError};

impl<S> AppService<S>
where
    S: PersistenceStore + Send,
{
    pub fn get_agent_runtime(
        &self,
        query: &GetAgentRuntimeQuery,
    ) -> Result<crate::AgentRuntimeSnapshot, AppServiceError> {
        self.agent_runtime
            .snapshot(query)
            .map_err(AppServiceError::from)
    }

    pub fn patch_agent_runtime_profile(
        &self,
        params: &DaemonAgentRuntimePatchProfileParams,
    ) -> Result<crate::AgentRuntimeSnapshot, AppServiceError> {
        self.agent_runtime
            .patch_profile(params)
            .map_err(AppServiceError::from)
    }

    pub async fn login_agent_runtime_auth_profile(
        &self,
        params: &DaemonAgentRuntimeAuthLoginParams,
    ) -> Result<AuthProfileLoginResult, AppServiceError> {
        self.agent_runtime
            .login_auth_profile(params)
            .await
            .map_err(AppServiceError::from)
    }

    pub fn replace_agent_runtime_auth_profile_preferences(
        &self,
        params: &DaemonAgentRuntimeAuthProfilePreferencesSetParams,
    ) -> Result<crate::AgentRuntimeSnapshot, AppServiceError> {
        self.agent_runtime
            .replace_auth_profile_preferences(params)
            .map_err(AppServiceError::from)
    }

    pub async fn logout_agent_runtime_auth_profile(
        &self,
        params: &DaemonAgentRuntimeAuthLogoutParams,
    ) -> Result<AuthProfileLogoutResult, AppServiceError> {
        self.agent_runtime
            .logout_auth_profile(params)
            .await
            .map_err(AppServiceError::from)
    }

    pub async fn complete_agent_runtime_auth_profile_login(
        &self,
        params: &DaemonAgentRuntimeAuthLoginCompleteParams,
    ) -> Result<AuthProfileLoginResult, AppServiceError> {
        self.agent_runtime
            .complete_auth_profile_login(params)
            .await
            .map_err(AppServiceError::from)
    }

    pub fn set_agent_runtime_extension_enabled(
        &self,
        params: &DaemonAgentRuntimeSetExtensionEnabledParams,
    ) -> Result<crate::AgentRuntimeSnapshot, AppServiceError> {
        self.agent_runtime
            .set_extension_enabled(params)
            .map_err(AppServiceError::from)
    }
}
